### PR TITLE
transports/quic: Add test

### PR DESCRIPTION
**Pull request by mxinden**

### CODE REVIEW

**Comment by mxinden** [comment]
@kpp I am not sure why the UI shows all the commits from `master`. I would expect it to show the merge commit only. I tested merging into your branch locally which works correctly. Maybe because your fork is not aware of the commits? If you synchronize your `master` branch with `libp2p/master` it might resolve?

**Comment by kpp** [comment]
I merged master. Can you rebase your code with tests?

**Comment by mxinden** [comment]
@kpp done :)

File: transports/quic/tests/smoke.rs [hunk 1]
```diff
@@ -23,7 +26,7 @@ async fn create_swarm(keylog: bool) -> Result<Swarm<RequestResponse<PingCodec>>>
     let mut transport = QuicConfig::new(keypair);
     transport
         .transport
-        .max_idle_timeout(Some(quinn_proto::VarInt::from_u32(1_000u32).into()));
+        .max_idle_timeout(Some(quinn_proto::VarInt::from_u32(60_000u32).into()));
```
Comment (kpp) [inline transports/quic/tests/smoke.rs:29]: Test `dial_failure` relies on timeouts:

```
test dial_failure has been running for over 60 seconds
test dial_failure ... ok
```

**Comment by kpp** [comment]
Nice test. Thank you!

File: transports/quic/tests/smoke.rs [hunk 2]
```diff
@@ -286,3 +290,118 @@ async fn dial_failure() -> Result<()> {
...
+        // For each listener node start `number_streams` requests.
+        for (listener_peer_id, listener_addr) in listeners {
+            dialer
+                .behaviour_mut()
+                .add_address(&listener_peer_id, listener_addr);
+
+            for _ in 0..number_streams {
+                dialer
+                    .behaviour_mut()
+                    .send_request(&listener_peer_id, Ping(data.clone()));
```
Comment (kpp) [inline transports/quic/tests/smoke.rs:372]: So I got a reproducible error with number_listeners=1, number_streams=2. And I debugged it a little bit.

* If number_streams=1, then:

<details><summary>CLICK ME</summary>

```
2022-02-21T22:19:16.471310Z  INFO prop{number_listeners=1 number_streams=1}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWFc25seNNmzjU31RcdQ8kybNnNmrY8tQvZ5EBsjereKbs")
2022-02-21T22:19:16.474056Z  INFO prop{number_listeners=1 number_streams=1}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWS78yehcDhQtA76jvekkFqnU7ZM517YTWzNBaqiDgApcv")
2022-02-21T22:19:16.474506Z  INFO prop{number_listeners=1 number_streams=1}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:19:16.474852Z  INFO prop{number_listeners=1 number_streams=1}: smoke: dialer Dialing
2022-02-21T22:19:16.477580Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:19:16.480826Z  INFO prop{number_listeners=1 number_streams=1}: smoke: listener IncomingConnection
2022-02-21T22:19:16.481264Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:19:16.481663Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:19:16.484304Z  INFO poll: libp2p_quic::upgrade: event=Ready(Connected)
2022-02-21T22:19:16.485011Z  INFO prop{number_listeners=1 number_streams=1}:inject_connection_established{peer=PeerId("12D3KooWFc25seNNmzjU31RcdQ8kybNnNmrY8tQvZ5EBsjereKbs") conn=ConnectionId(0) endpoint=Dialer { address: "/ip4/127.0.0.1/udp/50447/quic/p2p/12D3KooWFc25seNNmzjU31RcdQ8kybNnNmrY8tQvZ5EBsjereKbs" } _errors=Some([])}: libp2p_request_response: return=()
```
</details>
And test passes.<br /><br />

* If number_streams=2, then:

<details><summary>CLICK ME</summary>

```
2022-02-21T22:20:55.196925Z  INFO prop{number_listeners=1 number_streams=2}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8")
2022-02-21T22:20:55.199519Z  INFO prop{number_listeners=1 number_streams=2}:create_swarm{keylog=true}: smoke: peer_id=PeerId("12D3KooWF8yMitRvwRFNVkDYoMSQYFPR6JZPRnrfmhEaA7EmMhK9")
2022-02-21T22:20:55.199958Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:20:55.200070Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:20:55.200410Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer Dialing
2022-02-21T22:20:55.200576Z  INFO prop{number_listeners=1 number_streams=2}:inject_dial_failure{peer=Some(PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8")) _dial_error=DialPeerConditionFalse(Disconnected)}: libp2p_request_response: return=()
2022-02-21T22:20:55.200656Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer e=Some(Behaviour(OutboundFailure { peer: PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8"), request_id: RequestId(1), error: DialFailure }))
2022-02-21T22:20:55.200764Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer e=Some(Behaviour(OutboundFailure { peer: PeerId("12D3KooWSLuYD79Masc7vWnjBSjEub2wZfcvKa8nGBXvFVgbz4U8"), request_id: RequestId(2), error: DialFailure }))
2022-02-21T22:20:55.202648Z  INFO poll: libp2p_quic::upgrade: event=Pending
```
</details>

Dialing stops, test fails.<br /><br />

* But if number_streams=2 and I change `RequestResponse::send_request` `.condition(dial_opts::PeerCondition::Disconnected)`  to `PeerCondition::Always`, I got:

<details><summary>CLICK ME</summary>

```
2022-02-21T22:16:51.250703Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:16:51.250826Z  INFO prop{number_listeners=1 number_streams=2}:try_send_request: libp2p_request_response: peer not connected
2022-02-21T22:16:51.251160Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer Dialing
2022-02-21T22:16:51.251429Z  INFO prop{number_listeners=1 number_streams=2}: smoke: dialer Dialing
2022-02-21T22:16:51.253403Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.254043Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.256479Z  INFO prop{number_listeners=1 number_streams=2}: smoke: listener IncomingConnection
2022-02-21T22:16:51.256921Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.258330Z  INFO prop{number_listeners=1 number_streams=2}: smoke: listener IncomingConnection
2022-02-21T22:16:51.258388Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.258747Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.259084Z  INFO poll: libp2p_quic::upgrade: event=Pending
2022-02-21T22:16:51.261180Z  INFO poll: libp2p_quic::upgrade: event=Ready(Connected)
2022-02-21T22:16:51.262058Z  INFO prop{number_listeners=1 number_streams=2}:inject_connection_established{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE") conn=ConnectionId(0) endpoint=Dialer { address: "/ip4/127.0.0.1/udp/32856/quic/p2p/12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE" } _errors=Some([])}: libp2p_request_response: return=()
2022-02-21T22:16:51.262315Z  INFO prop{number_listeners=1 number_streams=2}:inject_connected{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE")}:try_send_request: libp2p_request_response: peer connected
2022-02-21T22:16:51.262422Z  INFO prop{number_listeners=1 number_streams=2}:inject_connected{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE")}:try_send_request: libp2p_request_response: peer connected
2022-02-21T22:16:51.262499Z  INFO prop{number_listeners=1 number_streams=2}:inject_connected{peer=PeerId("12D3KooWFMA9Zo7BJinx6Jm95N4rJcYDZkjbdQnj9iJXJGPQPLTE")}: libp2p_request_response: return=()
```
</details>

And test passes, even with number_listeners=5, number_streams=12.

Comment (kpp) [reply transports/quic/tests/smoke.rs:372]: @mxinden ping

Comment (mxinden) [reply transports/quic/tests/smoke.rs:372]: I am not able to make the test succeed either. I hope it is the test that is buggy, I suspect it is the implementation itself.

Comment (kpp) [reply transports/quic/tests/smoke.rs:372]: Here you are: https://github.com/kpp/rust-libp2p/tree/libp2p-quic-mxinden-test all good (except for ugh... disconnects, but I am working on it).

**Comment by kpp** [comment]
Fixed in https://github.com/kpp/rust-libp2p/pull/14

**Pull request closed without merging**